### PR TITLE
fix(lambda): grantInvoke on imported function fails

### LINKED ISSUE #10607: [elasticloadbalancingv2] Can't use imported lambda with ELBv2
*Description of the bug* Can't use imported lambda with ELBv2  

This is similar to [#8597](https://github.com/aws/aws-cdk/issues/8597) which was closed due to similar bugs in Lambda but ElasticLoadBalancingv2 module still needs to fix their part

### Reproduction Steps

This piece of code will produce the error below
```
    const lb = new elbv2.ApplicationLoadBalancer(this, 'LB', {
      vpc,
      internetFacing: false
    });

    const listener = lb.addListener('Listener', {
      port: 80,
    });

    const fn = lambda.Function.fromFunctionArn(this, 'myFunction', functionARN)

    listener.addTargets('Target', {
      port: 80,
      targets: [new targets.LambdaTarget(fn)]
    });
    
    // @aws-cdk/aws-elasticloadbalancingv2-targets.LambdaTarget
    const lambdaTarget = new targets.LambdaTarget(fn);
    
    new lambda.CfnPermission(this, 'MyRule_MyLambda_InvokePermission', {
          action: 'lambda:InvokeFunction',
          functionName: fn.functionArn,
          principal: 'elasticloadbalancing.amazonaws.com',
          sourceArn: listener.listenerArn
      })
    
    listener.addTargets('Target', {
      port: 80,
      targets: [lambdaTarget]
    })

    listener.connections.allowDefaultPortFromAnyIpv4('test');
```

### What actually happened?

when running *cdk synth* I get the following error
```
No child with id: 'InvokeServicePrincipal(elasticloadbalancing.amazonaws.com)'

/home/ec2-user/environment/monolithic-modernization-lab/root-endpoint/node_modules/@aws-cdk/core/lib/private/synthesis.ts:151
    throw new Error(`Validation failed with the following errors:\n  ${errorList}`);
          ^
Error: Validation failed with the following errors:
  [RootEndpointStack/LB/Listener] Listener needs at least one default action or target group (call addTargetGroups or addAction)
    at validateTree (/home/ec2-user/environment/monolithic-modernization-lab/root-endpoint/node_modules/@aws-cdk/core/lib/private/synthesis.ts:151:11)
    at Object.synthesize (/home/ec2-user/environment/monolithic-modernization-lab/root-endpoint/node_modules/@aws-cdk/core/lib/private/synthesis.ts:25:5)
    at App.synth (/home/ec2-user/environment/monolithic-modernization-lab/root-endpoint/node_modules/@aws-cdk/core/lib/stage.ts:175:23)
    at process.App.process.once (/home/ec2-user/environment/monolithic-modernization-lab/root-endpoint/node_modules/@aws-cdk/core/lib/app.ts:112:45)
    at Object.onceWrapper (events.js:286:20)
    at process.emit (events.js:198:13)
    at process.EventEmitter.emit (domain.js:448:20)
    at process.emit (/home/ec2-user/environment/monolithic-modernization-lab/root-endpoint/node_modules/source-map-support/source-map-support.js:495:21)
    at process.topLevelDomainCallback (domain.js:126:23)
```

### Environment

  - **CLI Version      : 1.64.1 (build 14a8c8d)**
  - **Framework Version: 1.64.1 **
  - **Node.js Version: v10.19.0** 
  - **OS               : Linux 4.14.165-103.209.amzn1.x86_64**
  - **Language (Version): TypeScript** 

### Other

Removing the lambda.CfnPermission produces the same issue.

Issue has been first reported in Bug Report [8597 ](https://github.com/aws/aws-cdk/issues/8597) - That report seems to be closed due to fixes in Lambda, but the aws-elasticloadbalancingv2 module still seems to be doing something else.

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by njlynch**
PR #8828 added the ability to add permissions on imported functions, if the
account on the stack and imported function matched. However, when this doesn't
match, attempting to call `grantInvoke` on the imported function results in a
`findChild` error.

This change guards against calling `addPermission` when `canCreatePermissions`
is false by checking if the `CfnPermission` was actually created, and adjusting
the return from `addInvoke` appropriately.

fixes #10607

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-lambda/test/test.lambda.ts [hunk 1]
```diff
@@ -1219,66 +1219,101 @@ export = {
...
+      test.done();
+    },
+
+    'on an imported function (different account)'(test: Test) {
+      // GIVEN
+      const stack = new cdk.Stack(); // Unresolved account
+      const fn = lambda.Function.fromFunctionArn(stack, 'Function', 'arn:aws:lambda:us-east-1:123456789012:function:MyFn');
+
+      // WHEN
+      fn.grantInvoke(new iam.ServicePrincipal('elasticloadbalancing.amazonaws.com'));
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-lambda/test/test.lambda.ts:1309]: Are we sure this shouldn't throw (as something that doesn't make sense?)

Can we add least add a `warning` to the construct tree or something?

Comment (njlynch) [reply packages/@aws-cdk/aws-lambda/test/test.lambda.ts:1309]: It's an unusual case, certainly. Effectively used by power users who already know their function has appropriate permissions and just need to add the reference to the target group(s). It certainly shouldn't fail with the current `findChild` error. Adding a warning to the tree makes sense.

File: packages/@aws-cdk/aws-lambda/lib/function-base.ts [hunk 2]
```diff
@@ -210,7 +210,7 @@ export abstract class FunctionBase extends Resource implements IFunction {
    */
   public addPermission(id: string, permission: Permission) {
     if (!this.canCreatePermissions) {
-      // FIXME: Report metadata
+      // FIXME: @deprecate this behavior in V2; throw an error if calling `addPermission` on a resource that doesn't support it.
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda/lib/function-base.ts:213]: use `@deprecated` as that would that be the search keyword

```suggestion
      // FIXME: @deprecated(v2) - throw an error if calling `addPermission` on a resource that doesn't support it.
```

File: packages/@aws-cdk/aws-lambda/lib/function-base.ts [hunk 3]
```diff
@@ -299,7 +299,11 @@ export abstract class FunctionBase extends Resource implements IFunction {
               action: 'lambda:InvokeFunction',
             });
 
-            return { statementAdded: true, policyDependable: this._functionNode().findChild(identifier) } as iam.AddToResourcePolicyResult;
+            const permissionNode = this._functionNode().tryFindChild(identifier);
+            if (!permissionNode) {
+              throw new Error('attempted to call `grantInvoke` on unsupported function (either imported or $LATEST version).');
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda/lib/function-base.ts:304]: Nit: something like this maybe, so it's a bit clearer to the user. The user might not know about 'grantInvoke'.

```suggestion
              throw new Error('Cannot modify permission to lambda function. Function is either imported or $LATEST version.');
```

**Review by nija-at — APPROVED** [review]

File: packages/@aws-cdk/aws-lambda/test/test.lambda.ts [hunk 4]
```diff
@@ -1220,66 +1223,118 @@ export = {
...
+    'on an imported function (different account)'(test: Test) {
+      // GIVEN
+      const stack = new cdk.Stack(undefined, undefined, {
+        env: { account: '111111111111' }, // Different account
+      });
+      const fn = lambda.Function.fromFunctionArn(stack, 'Function', 'arn:aws:lambda:us-east-1:123456789012:function:MyFn');
+
+      // THEN
+      test.throws(() => { fn.grantInvoke(new iam.ServicePrincipal('elasticloadbalancing.amazonaws.com')); },
+        'attempted to call `grantInvoke` on unsupported function (either imported or $LATEST version).');
```
Comment (njlynch) [inline packages/@aws-cdk/aws-lambda/test/test.lambda.ts:1334]:
```suggestion
        /Cannot modify permission to lambda function/);
```

**Pull request merged**